pkg/scheduler/eventhandlers/reservation_handler.go

```diff
@@ -0,0 +1,132 @@
+		AddFunc: func(obj interface{}) {
+			r, ok := obj.(*schedulingv1alpha1.Reservation)
+			if ok {
+				addReservation(sched, r)
+			} else {
+				klog.Errorf("reservation informer add func parse obj failed, obj %T", obj)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldR, oldOK := oldObj.(*schedulingv1alpha1.Reservation)
+			newR, newOK := newObj.(*schedulingv1alpha1.Reservation)
+			if !oldOK || !newOK {
+				klog.Errorf("unable to convert object to *schedulingv1alpha1.Reservation, old %T, new %T", oldObj, newObj)
+				return
+			}
+			updateReservation(sched, oldR, newR)
+		},
+		DeleteFunc: func(obj interface{}) {
+			switch t := obj.(type) {
+			case *schedulingv1alpha1.Reservation:
+				deleteReservation(sched, t)
+				return
+			case cache.DeletedFinalStateUnknown:
+				r, ok := t.Obj.(*schedulingv1alpha1.Reservation)
+				if ok {
+					deleteReservation(sched, r)
+					return
+				}
+			}
+			klog.Errorf("reservation informer delete func parse obj failed, obj %T", obj)
+		},
+	})
+}
+
+func addReservation(sched *scheduler.Scheduler, r *schedulingv1alpha1.Reservation) {
+	if r == nil || r.Spec.Template == nil {
+		klog.Errorf("failed to add reservation for invalid reservation %v", r)
+		return
+	}
+	reservePod := reservation.GetReservePod(r)
+	// TBD: should the reservation with nodeName specified enqueue?
+	if reservePod.Status.Phase != schedulingv1alpha1.ReasonReservationExpired {
+		err := sched.SchedulingQueue.Add(reservePod)
+		if err != nil {
+			klog.Errorf("failed to add reserve pod into scheduling queue, reservation %v, err: %v", reservation.GetReservationNamespacedName(r), err)
+		}
+		klog.V(4).Infof("add reservation %v", reservation.GetReservationNamespacedName(r))
+	}
+}
+
+func updateReservation(sched *scheduler.Scheduler, oldR, newR *schedulingv1alpha1.Reservation) {
+	if oldR == nil || newR == nil {
+		klog.Errorf("failed to update reservation for invalid reservation old %v, new %v", oldR, newR)
+		return
+	}
+	// only react to status.phase change
+	if oldR.Status.Phase == newR.Status.Phase {
+		klog.V(5).Infof("skip for status.phase of reservation %v not change", reservation.GetReservationNamespacedName(newR))
+		return
+	}
+	klog.V(4).Infof("got reservation %v phase update, old %v, new %v", reservation.GetReservationNamespacedName(newR), oldR.Status.Phase, newR.Status.Phase)
+
+	if oldR.Status.Phase == schedulingv1alpha1.ReservationAvailable &&
+		newR.Status.Phase == schedulingv1alpha1.ReservationExpired {
+		deleteReservation(sched, newR)
+	}
+}
+
+func deleteReservation(sched *scheduler.Scheduler, r *schedulingv1alpha1.Reservation) {
+	if r == nil || r.Spec.Template == nil {
+		klog.Errorf("failed to delete reservation for invalid reservation %v", r)
+		return
+	}
+
+	reservePod := reservation.GetReservePod(r)
+	// if reservation is scheduled, delete it from the schedule cache;
+	// otherwise, remove it from the scheduling queue.
+	if reservation.IsReservationScheduled(r) {
```

Comment:
Could there be a concurrency race condition problem? When deleting the Reservation, the ReservePod may be scheduled exactly.

Reply:
fixed